src/posix/platform/netif.cpp

```diff
@@ -505,6 +513,202 @@ static void UpdateLink(otInstance *aInstance)
+
+    VerifyOrExit(ifIdx > 0, error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(sNetlinkFd >= 0, error = OT_ERROR_INVALID_STATE);
+
+    req.header.nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK | NLM_F_NONREC;
+
+    req.header.nlmsg_len  = NLMSG_LENGTH(sizeof(rtmsg));
+    req.header.nlmsg_type = RTM_DELROUTE;
+    req.header.nlmsg_pid  = 0;
+    req.header.nlmsg_seq  = ++sNetlinkSequence;
+
+    req.msg.rtm_family   = AF_INET6;
+    req.msg.rtm_src_len  = 0;
+    req.msg.rtm_dst_len  = aPrefix.mLength;
+    req.msg.rtm_tos      = 0;
+    req.msg.rtm_scope    = RT_SCOPE_UNIVERSE;
+    req.msg.rtm_type     = RTN_UNICAST;
+    req.msg.rtm_table    = RT_TABLE_MAIN;
+    req.msg.rtm_protocol = RTPROT_BOOT;
+    req.msg.rtm_flags    = 0;
+
+    otIp6AddressToString(&aPrefix.mPrefix, addrBuf, OT_IP6_ADDRESS_STRING_SIZE);
+    inet_pton(AF_INET6, addrBuf, data);
+    SuccessOrExit(error = AddRtAttr(&req.header, sizeof(req), RTA_DST, data, sizeof(data)));
+    SuccessOrExit(error = AddRtAttr32(&req.header, sizeof(req), RTA_OIF, ifIdx));
+
+    VerifyOrExit(send(sNetlinkFd, &req, sizeof(req), 0) > 0, error = OT_ERROR_FAILED);
+
+exit:
+    return error;
+}
+
+bool HasExternalRouteInNetData(otInstance *aInstance, const otIp6Prefix &aExternalRoute)
+{
+    otNetworkDataIterator iterator = OT_NETWORK_DATA_ITERATOR_INIT;
+    otExternalRouteConfig config;
+    bool                  found = false;
+
+    while (otNetDataGetNextRoute(aInstance, &iterator, &config) == OT_ERROR_NONE)
+    {
+        if (otIp6IsAddressEqual(&config.mPrefix.mPrefix, &aExternalRoute.mPrefix) &&
+            config.mPrefix.mLength == aExternalRoute.mLength)
+        {
+            found = true;
+            break;
+        }
+    }
+    return found;
+}
+
+bool HasAddedExternalRoute(const otIp6Prefix &aExternalRoute)
+{
+    bool found = false;
+
+    for (int i = 0; i < static_cast<int>(sAddedExternalRoutesNum); ++i)
+    {
+        if (otIp6IsAddressEqual(&sAddedExternalRoutes[i].mPrefix, &aExternalRoute.mPrefix) &&
+            sAddedExternalRoutes[i].mLength == aExternalRoute.mLength)
+        {
+            found = true;
+            break;
+        }
+    }
+    return found;
+}
+
+static void UpdateExternalRoutes(otInstance *aInstance)
+{
+    otError               error;
+    otNetworkDataIterator iterator = OT_NETWORK_DATA_ITERATOR_INIT;
+    otExternalRouteConfig config;
+
+    for (int i = 0; i < static_cast<int>(sAddedExternalRoutesNum); ++i)
+    {
+        if (HasExternalRouteInNetData(aInstance, sAddedExternalRoutes[i]))
+        {
+            continue;
+        }
+        if ((error = DeleteExternalRoute(sAddedExternalRoutes[i])) != OT_ERROR_NONE)
+        {
+            otLogWarnPlat("failed to delete an external route in kernel: %s", otThreadErrorToString(error));
+        }
+        else
+        {
+            sAddedExternalRoutes[i] = sAddedExternalRoutes[sAddedExternalRoutesNum - 1];
+            --sAddedExternalRoutesNum;
+            --i;
+        }
```

Comment:
```suggestion
        if ((error = DeleteExternalRoute(sAddedExternalRoutes[i])) != OT_ERROR_NONE)
        {
            otLogWarnPlat("failed to delete an external route in kernel: %s", otThreadErrorToString(error));
        }

        sAddedExternalRoutes[i] = sAddedExternalRoutes[sAddedExternalRoutesNum - 1];
        --sAddedExternalRoutesNum;
        --i;
```
I think we can continue to remove the route from `sAddedExternalRoutes` even when we failed to remove it from kernel, since it's unlikely that retrying will be successful.

Reply:
If we removes the route from `sAddedExternalRoutes` then we won't be able to track and remove it from kernel before rebooting. By the way, why is retrying unlike to succeed?

Reply:
Fine to keep it as is.